src-input/duk_bi_encoding.c

```diff
@@ -115,34 +131,53 @@ DUK_LOCAL void duk__utf8_encode_char(void *udata, duk_codepoint_t codepoint) {
 	duk__encode_context *enc_ctx;
 
 	enc_ctx = (duk__encode_context *) udata;
-	if (codepoint > 0x10ffffL) {
+
+#if !defined(DUK_USE_PREFER_SIZE)
+	if (codepoint <= 0x7f && enc_ctx->lead == 0x0000L) {
+		/* Fast path for ASCII. */
+		*enc_ctx->out++ = (duk_uint8_t) codepoint;
+		return;
+	}
+#endif
+
+	if (DUK_UNLIKELY(codepoint > 0x10ffffL)) {
 		/* cannot legally encode in UTF-8 */
-		enc_ctx->out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx->out);
-	} else if (codepoint >= 0xd800L && codepoint <= 0xdbffL) {
-		/* high surrogate */
-		if (enc_ctx->lead != 0x0000L) {
-			/* consecutive high surrogates, consider first one unpaired */
-			enc_ctx->out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx->out);
-		}
-		enc_ctx->lead = codepoint;
-	} else if (codepoint >= 0xdc00L && codepoint <= 0xdfffL) {
-		/* low surrogate */
-		if (enc_ctx->lead != 0x0000L) {
-			codepoint = 0x010000L | ((enc_ctx->lead - 0xd800L) << 10) | (codepoint - 0xdc00L);
-			enc_ctx->out += duk_unicode_encode_xutf8(codepoint, enc_ctx->out);
-			enc_ctx->lead = 0x0000L;
+		codepoint = DUK_UNICODE_CP_REPLACEMENT_CHARACTER;
+	} else if (codepoint >= 0xd800L && codepoint <= 0xdfffL) {
+		if (codepoint <= 0xdbffL) {
+			/* high surrogate */
+			duk_codepoint_t prev_lead = enc_ctx->lead;
+			enc_ctx->lead = codepoint;
+			if (prev_lead == 0x0000L) {
+				/* high surrogate, no output */
+				return;
+			} else {
+				/* consecutive high surrogates, consider first one unpaired */
+				codepoint = DUK_UNICODE_CP_REPLACEMENT_CHARACTER;
+			}
 		} else {
-			/* unpaired low surrogate */
-			enc_ctx->out += duk_unicode_encode_xutf8(DUK_UNICODE_CP_REPLACEMENT_CHARACTER, enc_ctx->out);
+			/* low surrogate */
+			if (enc_ctx->lead != 0x0000L) {
+				codepoint = 0x010000L + ((enc_ctx->lead - 0xd800L) << 10) + (codepoint - 0xdc00L);
```

Comment:
A quick test program shows the difference; with bitwise ops:

```
  400461:       2d 00 d8 00 00          sub    $0xd800,%eax
  400466:       81 ee 00 dc 00 00       sub    $0xdc00,%esi
  40046c:       c1 e0 0a                shl    $0xa,%eax
  40046f:       81 ce 00 00 01 00       or     $0x10000,%esi
  400475:       09 c6                   or     %eax,%esi
```

With addition:

```
  400461:       c1 e0 0a                shl    $0xa,%eax
  400464:       8d b4 02 00 24 a0 fc    lea    -0x35fdc00(%rdx,%rax,1),%esi
```

The key issue is that with addition the addition and subtractions can be precomputed. With OR the components themselves involve subtraction which doesn't combine well during compile time.